op-e2e/actions/proofs/isthmus_setcode_tx_test.go

```diff
@@ -189,3 +196,224 @@ func testInvalidSetCodeTxBatch(gt *testing.T, testCfg *helpers.TestCfg[any]) {
+
+	// read prefix
+	_, err = binary.ReadUvarint(batchBuf)
+	require.NoError(t, err, "failed to read rel timestamp")
+	_, err = binary.ReadUvarint(batchBuf)
+	require.NoError(t, err, "failed to read l1 origin num timestamp")
+	_, err = batchBuf.Read(make([]byte, 40))
+	require.NoError(t, err, "failed to read parent and origin checks")
+
+	// read payload
+	blockCount, err := binary.ReadUvarint(batchBuf)
+	require.NoError(t, err, "failed to read block count")
+
+	originBitfieldSize := (blockCount + 7) / 8
+	originBitfield := make([]byte, originBitfieldSize)
+	_, err = batchBuf.Read(originBitfield)
+	require.NoError(t, err, "failed to read origin bitfield")
+
+	blockTxCounts := make([]uint64, blockCount)
+	totalBlockTxCount := uint64(0)
+
+	for i := uint64(0); i < blockCount; i++ {
+		blockTxCount, err := binary.ReadUvarint(batchBuf)
+		require.NoError(t, err, "failed to read block tx count")
+
+		blockTxCounts[i] = blockTxCount
+		totalBlockTxCount += blockTxCount
+	}
+
+	// read contract creation bits
+	contractCreationBits := make([]byte, (totalBlockTxCount+7)/8)
+	_, err = batchBuf.Read(contractCreationBits)
+	require.NoError(t, err, "failed to read contract creation bits")
+
+	offsetStart := uint64(batchBuf.CurrPos()) - uint64(totalBlockTxCount+7)/8 + 1
+	require.Equal(t, contractCreationBits[0], byte(0b10), "expected contract creation bits to be 0b10")
+
+	// flip the second bit (big endian)
+	rlpDecoded[offsetStart] = 0b01 // unset bit 2 and set bit 1
+
+	// re-encode rlp
+	var buf bytes.Buffer
+	err = rlp.Encode(&buf, rlpDecoded)
+	require.NoError(t, err)
+
+	// re-compress
+	compressedBuf := new(bytes.Buffer)
+	compressor, err := zlib.NewWriterLevel(compressedBuf, zlib.BestCompression)
+	require.NoError(t, err)
+	_, err = compressor.Write(buf.Bytes())
+	require.NoError(t, err)
+	err = compressor.Close()
+	require.NoError(t, err)
+
+	// replace the frame data with the new compressed data
+	newOutputFrame := make([]byte, 1+16+2+4+len(compressedBuf.Bytes())+1)
+	copy(newOutputFrame[:1+16+2], outputFrame[:1+16+2])
+	binary.BigEndian.PutUint32(newOutputFrame[1+16+2:1+16+2+4], uint32(len(compressedBuf.Bytes())))
+	copy(newOutputFrame[1+16+2+4:], compressedBuf.Bytes())
+	newOutputFrame[len(newOutputFrame)-1] = outputFrame[len(outputFrame)-1]
+
+	// submit new batch
+	batcher.ActL2BatchSubmitRaw(t, newOutputFrame)
+
+	miner.ActL1StartBlock(12)(t)
+	miner.ActL1IncludeTxByHash(env.Batcher.LastSubmitted.Hash())(t)
+	miner.ActL1EndBlock(t)
+
+	sequencer.ActL1HeadSignal(t)
+	sequencer.ActL2PipelineFull(t)
+
+	l2safe := sequencer.L2Safe()
+	s2block := env.Engine.L2Chain().GetBlockByHash(l2safe.Hash)
+	require.Len(t, s2block.Transactions(), 0, "safe head should not contain either tx")
+	require.Equal(t, u1, l2safe, "expected last block to be reorgd out due to setcode tx")
+
+	// recs := env.Logs.FindLogs(testlog.NewMessageFilter("to address is required for SetCodeTx"))
+	// require.Len(t, recs, 1)
```

Comment:
The bit twiddling magic above is impressive but also looks quite complex and error prone 😅  I think it's essential that we find the right log, or we don't have confirmation that we cover here what we want.

I would prefer for us to go a different route and add a testing mode to a span batch (maybe a wrapper type `TestSpanBatch` around a span batch that has additional methods that we can use in tests) so that we can manipulate the contract creation bits conveniently via methods. And we could then add some form of injection point to the action batcher to modify the span batch before compressing it into a channel.

Reply:
I generally agree, although I wasn't sure if the injection route would be much less code. The log is produced, but for some reason, this doesn't pick it up. We'll have to add a few layers of injection to get this working if we aren't modifying the span batch directly. Overall, I don't think it's a horrible idea to create a test util to modify a compressed span batch. Also, span batch format must stay backwards compatible anyway, so the only issue would be if we changed the default format for tests, and it should be relatively easy to track down any failures.

Reply:
I attempted to implement the hooks [here](https://github.com/ethereum-optimism/optimism/pull/14928/commits/942e7c6a55f2778bc1485627feb378be343ef53c), but it required a ton of refactoring. I had to make fields public on SpanBatchTxs and RawSpanBatch so I could modify them in the hook. 

Also had to update `Buffer()` to take in channel creation options. Making the fields public allows much easier decoding/encoding externally (would help avoid copying all of the serialization/deserialization code I had to write to just modify a field in a span batch), so I think this change does make sense.

I can also try splitting up this change if that's helpful.